### PR TITLE
Allow For Proper Registration and Deregistration of Topics

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Altair Hard Fork

**What does this PR do? Why is it needed?**

- [x] Allows for old gossip topics to be de-registered.
- [x] Prevents duplicate registration of topics.
- [x] Properly cleanup relevant subnet topics from older forks for the attestation 
and sync subnets.
 
**Which issues(s) does this PR fix?**

N.A

**Other notes for review**

### CODE REVIEW

File: beacon-chain/sync/fork_watcher.go [hunk 1]
```diff
@@ -32,10 +32,60 @@ func (s *Service) forkWatcher() {
...
+						}
+					}
+					if digestExists {
+						continue
+					}
 					s.registerSubscribers(nextEpoch, digest)
 					s.registerRPCHandlersAltair()
 				}
 			}
+			// This routine takes care of the de-registration of
```
Comment (terencechain) [inline beacon-chain/sync/fork_watcher.go:54]: Tests for this routine?

File: beacon-chain/sync/subscriber.go [hunk 2]
```diff
@@ -456,6 +479,24 @@ func (s *Service) subscribeStaticWithSyncSubnets(topic string, validator pubsub.
 				if s.chainStarted.IsSet() && s.cfg.InitialSync.Syncing() {
 					continue
 				}
+				valid, err := isDigestValid(digest, genesis, genRoot)
```
Comment (terencechain) [inline beacon-chain/sync/subscriber.go:482]: Tests for here as well

File: beacon-chain/sync/fork_watcher.go [hunk 3]
```diff
@@ -32,10 +32,60 @@ func (s *Service) forkWatcher() {
...
+							continue
+						}
+						if retDigest == digest {
+							digestExists = true
+							break
+						}
+					}
+					if digestExists {
+						continue
+					}
```
Comment (prestonvanloon) [inline beacon-chain/sync/fork_watcher.go:49]: Can we create a map for this? O(n) lookup is not great.

File: beacon-chain/sync/fork_watcher.go [hunk 4]
```diff
@@ -32,10 +32,60 @@ func (s *Service) forkWatcher() {
...
+			}
+			epochAfterFork := currFork.Epoch + 1
+			nonGenesisFork := currFork.Epoch > 1
+			// If we are in the epoch after the fork, we start de-registering.
+			if epochAfterFork == currEpoch && nonGenesisFork {
+				// Look at the previous fork's digest.
+				epochBeforeFork := currFork.Epoch - 1
+				prevDigest, err := p2putils.ForkDigestFromEpoch(epochBeforeFork, genRoot[:])
+				if err != nil {
+					log.WithError(err)
```
Comment (prestonvanloon) [inline beacon-chain/sync/fork_watcher.go:70]: The `WithError` method is a helper to add a field to the log. You still need to log something here. 
```suggestion
					log.WithError(err).Error("Failed to determine previous epoch fork digest")
```

File: beacon-chain/sync/fork_watcher.go [hunk 5]
```diff
@@ -32,10 +32,60 @@ func (s *Service) forkWatcher() {
...
+					continue
+				}
+				// Run through all our current active topics and see
+				// if there are any subscriptions to be removed.
+				for _, t := range s.subHandler.allTopics() {
+					retDigest, err := digestFromTopic(t)
```
Comment (prestonvanloon) [inline beacon-chain/sync/fork_watcher.go:76]: The digest map would help here too.

File: beacon-chain/sync/subscription_topic_handler.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
+package sync
```
Comment (prestonvanloon) [inline beacon-chain/sync/subscription_topic_handler.go:1]: Please add tests for subscription topic handler

**Review by prestonvanloon — CHANGES_REQUESTED** [review]

File: beacon-chain/sync/fork_watcher.go (hunk 3 above)
Comment (nisdas) [reply beacon-chain/sync/fork_watcher.go]: Done, refactored it to be handled in the subscription handler.

File: beacon-chain/sync/subscriber.go (hunk 2 above)
Comment (nisdas) [reply beacon-chain/sync/subscriber.go:481]: This whole method is difficult to measure as it is ticker based, so race conditions are abound. Also it involves tricky logic with libp2p for testing. I want to refactor away this whole file later so testing is easier to carry out here.

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**